File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        // if (root == null && subRoot != null ) return false;
        if (root == null ) return false;
        if (sameTree(root,subRoot)) return true;

        Boolean v1 = false;
        Boolean v2 = false;
        if (!sameTree(root,subRoot)) {
            v1 =  isSubtree(root.left, subRoot);
            v2 = isSubtree(root.right, subRoot);

        }
           return (v1 || v2);
    }

    public boolean sameTree(TreeNode root, TreeNode subRoot) {
        if (root == null && subRoot == null) {
            return true;
        }
        if (root != null && subRoot != null && root.val == subRoot.val) {
            return sameTree(root.left, subRoot.left) &&
                   sameTree(root.right, subRoot.right);
        }
        return false;
    }
}

File: Data Structures & Algorithms/binary-tree-zigzag-level-order-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return []

        result = []

        queue = deque([root])

        switch = True
        while queue:
            level_size = len(queue)
            curr_level = []

            for _ in range(level_size):
                node = queue.popleft()
                curr_level.append(node.val)

                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            if switch:
                result.append(curr_level)
                switch = not switch
            else:
                result.append(curr_level[::-1])
                switch = not switch


        return result